src/python/grpcio_tests/tests/health_check/_health_servicer_test.py

```diff
@@ -203,6 +203,30 @@ def test_cancelled_watch_removed_from_watch_list(self):
                 'watch set should be empty')
             self.assertTrue(response_queue.empty())
 
+        def test_graceful_shutdown(self):
+            request = health_pb2.HealthCheckRequest(service='')
+            response_queue = queue.Queue()
+            rendezvous = self._stub.Watch(request)
+            thread = threading.Thread(
+                target=_consume_responses, args=(rendezvous, response_queue))
+            thread.start()
+
+            response = response_queue.get(timeout=test_constants.SHORT_TIMEOUT)
+            self.assertEqual(health_pb2.HealthCheckResponse.SERVING,
+                             response.status)
+
+            self._servicer.enter_graceful_shutdown()
+            response = response_queue.get(timeout=test_constants.SHORT_TIMEOUT)
+            self.assertEqual(health_pb2.HealthCheckResponse.NOT_SERVING,
+                             response.status)
+
+            # This should be a no-op.
+            self._servicer.set('', health_pb2.HealthCheckResponse.SERVING)
```

Comment:
Should we check if the `response_queue` is not getting any new response?

Reply:
This is done below after cancelling the call and joining the thread. This may be susceptible to false-negatives if the server response can race with the cancellation (I'm not sure how best to check for something *not* happening) but I "verify" manually that this check fails if the `self._gracefully_shutting_down` check is removed from `HealthServicer#set`.